biogeochem/FatesLandUseChangeMod.F90

```diff
@@ -0,0 +1,307 @@
+
+  subroutine get_luh_statedata(bc_in, state_vector)
+
+    type(bc_in_type) , intent(in) :: bc_in
+    real(r8), intent(out) :: state_vector(n_landuse_cats)  ! [m2/m2]
+
+    ! LOCALS
+    type(luh2_fates_lutype_map) :: lumap
+    real(r8) :: temp_vector(hlm_num_luh2_states)  ! [m2/m2]
+    real(r8) :: urban_fraction
+    integer  :: i_luh2_states
+    integer  :: ii
+    character(5) :: state_name
+    logical :: modified_flag
+
+    ! zero state vector and urban fraction
+    state_vector(:) = 0._r8
+    urban_fraction = 0._r8
+
+    ! Check to see if the incoming state vector is NaN.
+    temp_vector = bc_in%hlm_luh_states
+    call CheckLUHData(temp_vector,modified_flag)
+    if (.not. modified_flag) then
+       ! identify urban fraction so that it can be factored into the land use state output
+       urban_fraction = bc_in%hlm_luh_states(findloc(bc_in%hlm_luh_state_names,'urban',dim=1))
+    end if
+
+    ! loop over all states and add up the ones that correspond to a given fates land use type
+    do i_luh2_states = 1, hlm_num_luh2_states
+
+       ! Get the luh2 state name and determine fates aggregated land use
+       ! type index from the state to lutype map
+       state_name = bc_in%hlm_luh_state_names(i_luh2_states)
+       ii = lumap%GetIndex(state_name)
+
+       ! Avoid 'urban' states whose indices have been given unset values
+       if (ii .ne. fates_unset_int) then
+          state_vector(ii) = state_vector(ii) + &
+               temp_vector(i_luh2_states) / (1._r8 - urban_fraction)
+       end if
+    end do
+
+    ! check to ensure total area == 1, and correct if not
+    if ( abs(sum(state_vector(:)) - 1._r8) .gt. nearzero ) then
+       write(fates_log(),*) 'warning: sum(state_vector) = ', sum(state_vector(:))
+       state_vector = state_vector / sum(state_vector)
+    end if
+
+  end subroutine get_luh_statedata
+
+  !----------------------------------------------------------------------------------------------------
+
+  subroutine CheckLUHData(luh_vector,modified_flag)
+
+    use shr_infnan_mod   , only : isnan => shr_infnan_isnan
+
+    real(r8), intent(inout) :: luh_vector(:)  ! [m2/m2]
+    logical, intent(out)    :: modified_flag
+
+    ! Check to see if the incoming luh2 vector is NaN.
+    ! This suggests that there is a discepency where the HLM and LUH2 states
+    ! there is vegetated ground. E.g. LUH2 data is missing for glacier-margin regions such as Antarctica.
+    ! In this case, states should be Nan.  If so,
+    ! set the current state to be all primary forest, and all transitions to be zero.
+    ! If only a portion of the vector is NaN, there is something  amiss with
+    ! the data, so end the run.
+
+    modified_flag = .false.
+    if (all(isnan(luh_vector))) then
+       luh_vector(:) = 0._r8
+       ! Check if this is a state vector, otherwise leave transitions as zero
+       if (size(luh_vector) .eq. hlm_num_luh2_states) then
```

Comment:
Checked this through, considered what to do with states versus transitions, and I follow your logic and the indexing.